File: leetcode/00095_unique-binary-search-trees-ii/95-unique-binary-search-trees-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// A hash function used to hash a tuple
struct hash_tuple {
    template <class T1, class T2>
  
    size_t operator()(const tuple<T1, T2> &x) const {
        return get<0>(x) ^ get<1>(x);
    }
};
 
class Solution {
public: 
    unordered_map<tuple<int, int>, vector<TreeNode*>, hash_tuple> cache;
    vector<TreeNode*> generateTrees(int n) {
        if (n == 0) return vector<TreeNode*>();
        return dfs(1, n);
    }
    
    vector<TreeNode*> dfs(int start, int end) {
        vector<TreeNode*> tree_list;
        if (start > end) {
            tree_list.push_back(NULL);
            return tree_list;
        }
        
        if(cache.count({start, end}) > 0) {
            return cache[{start, end}];
        }
        
        for (int i = start; i <= end; i++) {
            vector<TreeNode*> left = dfs(start, i-1);
            vector<TreeNode*> right = dfs(i + 1, end);
            for (TreeNode *l : left) {
                for (TreeNode *r : right) {
                    TreeNode *root = new TreeNode(i);
                    root->left = l;
                    root->right = r;
                    tree_list.push_back(root);
                }
            }
        }
        
        cache[{start, end}] = tree_list;
        return tree_list;
    }
};

File: leetcode/00095_unique-binary-search-trees-ii/95-unique-binary-search-trees-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        self.cache = {}
        return self.dfs(1, n)
    
    def dfs(self, start, end):
        if start > end:
            return [None]
        if (start, end) in self.cache: 
            return self.cache[(start, end)]               
        
        tree_list = []
        for idx in range(start, end + 1):
            left_child = self.dfs(start, idx - 1)
            right_child = self.dfs(idx + 1, end)
            for left in left_child:
                for right in right_child:
                    root = TreeNode(idx)
                    root.left = left
                    root.right = right
                    tree_list.append(root)
        self.cache[(start, end)] = tree_list
        
        return tree_list
 